tailfs/compositefs/openfile.go

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package compositefs
+
+import (
+	"context"
+	"io/fs"
+	"os"
+
+	"github.com/tailscale/xnet/webdav"
+	"tailscale.com/tailfs/shared"
+)
+
+// OpenFile implements interface webdav.Filesystem.
+func (cfs *compositeFileSystem) OpenFile(ctx context.Context, name string, flag int, perm os.FileMode) (webdav.File, error) {
+	if !shared.IsRoot(name) {
+		pathInfo, err := cfs.getPathInfo(name)
+		if err != nil {
+			return nil, err
+		}
+
+		if pathInfo.refersToChild {
+			// this is the child itself, ask it to open its root
+			return pathInfo.child.FS.OpenFile(ctx, "/", flag, perm)
+		}
+
+		return pathInfo.child.FS.OpenFile(ctx, pathInfo.pathOnChild, flag, perm)
+	}
+
+	// the root directory contains one directory for each child
+	di, err := cfs.Stat(ctx, name)
+	if err != nil {
+		return nil, err
+	}
+
+	return &shared.DirFile{
+		Info: di,
+		LoadChildren: func() ([]fs.FileInfo, error) {
+			cfs.childrenMu.Lock()
+			children := cfs.children
+			cfs.childrenMu.Unlock()
+
+			childInfos := make([]fs.FileInfo, 0, len(cfs.children))
+			for _, c := range children {
+				if c.isAvailable() {
+					var childInfo fs.FileInfo
+					if cfs.statChildren {
+						fi, err := c.FS.Stat(ctx, "/")
+						if err != nil {
+							return nil, err
+						}
+						// we use the full name
+						childInfo = shared.RenamedFileInfo(ctx, c.Name, fi)
```

Comment:
We could avoid the need for an intermediate by appending directly in these locations:
```go
   // ...
   childInfos = append(childInfos, shared.RenamedFileInfo(ctx, c.Name, fi))
} else {
   childInfos = append(childInfos, shared.ReadOnlyDirInfo(c.Name, cfs.now())
}

Reply:
Yes, but then we're repeating ourselves. 6 of one, half dozen of the other to me.

Reply:
Right, but it means the reader doesn't have to go look for where the variable is declared and reason about whether it's assigned on all paths. It's not repeating anything, those are two _different_ appends.